web/source/dom/domEventHandlers.ts

```diff
@@ -734,6 +675,156 @@ namespace com.keyman.dom {
+        //
+        // To properly reuse the InputHTMLElement-based x-coord search, we need the bounds to be TIGHT after
+        // this comment's containing block finishes. As in, we need [cpMin, cpMax] === [start of row, end of row].
+        //
+        // Note:  in certain scenarios, one of the bounds can be VERY far away at this point.
+        // Say, if a new touch event wants to move the caret nearby the previous location, as the
+        // bound on the opposite side may not have gotten overwritten even once so far in this call!
+        // Therefore... more binary search.
+
+        // Determine minimum caret position on the target row.
+        let minCpMin = cpMin; // Is before the row.
+        let maxCpMin = cp;
+        while(maxCpMin - minCpMin > 1) {
+          const searchPt = Math.round((maxCpMin+minCpMin)/2);
+          target.setTextCaret(searchPt);
+
+          const y=dom.Utils.getAbsoluteY(caret)-dy;  //top of caret
+          if(y < minY) {
+            minCpMin = searchPt+1; // We already know it's not on this row, so the minimum possible should be increased.
+          } else {
+            maxCpMin = searchPt; // Still in the same row, so the boundary can only be at or before this point.
+          }
+          cpMin = searchPt;
+        }
+
+        // Determine maximum caret position on the target row.
+        let minCpMax = cp;
+        let maxCpMax = cpMax; // Is after the row.
+        while(maxCpMax - minCpMax > 1) {
+          const searchPt = Math.round((maxCpMax+minCpMax)/2);
+          target.setTextCaret(searchPt);
+
+          const y=dom.Utils.getAbsoluteY(caret)-dy;  //top of caret
+          if(y > maxY) {
+            maxCpMax = searchPt-1; // We already know it's not on this row, so the maximum possible should be decreased.
+          } else {
+            minCpMax = searchPt; // Still in the same row, so the boundary can only be at or after this point.
+          }
+          cpMax = searchPt;
+        }
+
+        // Set the potential caret in the middle of the range.
+        cp = Math.round((cpMin + cpMax)/2);
+        target.setTextCaret(cp);
+      }
+
+      // Caret repositioning for horizontal scrolling of RTL text
+
+      // snapOrder - 'snaps' the touch location in a manner corresponding to the 'ltr' vs 'rtl' orientation.
+      // Think of it as performing a floor() function, but the floor depends on the origin's direction.
+      var snapOrder;
+      if(target.dir == 'rtl') {  // I would use arrow functions, but IE doesn't like 'em.
+        snapOrder = function(a, b) {
+          return a < b;
+        };
+      } else {
+        snapOrder = function(a, b) {
+          return a > b;
+        };
+      }
+
+      // Now to binary-search the x-coordinate.
+      // Except... we... haven't modified cpMin & cpMax since their roles in the
+      // y-position search, which matters if the backing element is a TextArea.
+      // Why not?  That sounds dangerous!
```

Comment:
I have no idea what this comment means?!

Reply:
That comment was written during my analysis of the old code; there was faulty logic when repurposing the `<input>`-oriented horizontal positioning code, which assumes `cpMin` and `cpMax` to be boundary positions on the target line.  The `<textarea>` branch never tightened the bounds.

I can probably clean that up now that the bug's removed, but it was a critical note when diagnosing the caret-misplacement aspect noted in the description:

> While I was at it, I also fixed up the caret placement logic; it should land where expected far, far more consistently and reliably than before.

Reply:
Yes, please clean up -- comments are intended for future readers who don't know the history or context. So write them accordingly :grin: